File: ABC_250to300/ABC_270/C-SimplePath.rb
n, x, y = gets.chomp.split.map(&:to_i)
x -= 1
y -= 1
tree = Array.new(n).map{Array.new}

(n - 1).times do |i|
    u, v = gets.chomp.split.map{|hoge| hoge.to_i - 1}
    tree[u] << v;
    tree[v] << u;
end

$hash = {}
def dfs(index, y, route, tree) 
    #puts "index:#{index} route:#{route.join(' ')}"
    route << index 
    $hash[index] = true

    if index == y
        puts route.map{|r| r + 1}.join(" ")
        exit(0)
    end

    tree[index].each do |i|
        if $hash.has_key?(i) == true
            next
        end 

        dfs(i, y, route, tree)
    end

    route.pop
    return route
end

route= [] 
dfs(x, y, route, tree)

File: ABC_250to300/ABC_270/C-SimplePath.cpp
#include <algorithm>
#include <cmath>
#include <functional>
#include <iomanip>
#include <iostream>
#include <limits>
#include <map>
#include <numeric>
#include <queue>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;
const double PI = 3.141592653589793;
#define rep(i, n) for (int i = 0; i < (int)(n); i++)
typedef long long int              ll;
typedef vector<ll>                 vll;
typedef vector<vector<ll>>         vvll;
typedef vector<vector<vector<ll>>> vvvll;

unordered_map<ll, bool> dic;

stack<ll> dfs(ll index, ll y, stack<ll> &route, const vvll &tree) {
    // cout << "index:" << index << endl;

    route.push(index);
    dic[index] = true;

    if (index == y) {
        string str = "";
        while (route.empty() == false) {
            str += to_string(route.top() + 1);
            route.pop();
        }
        for (int i = str.size() - 1; i >= 0; --i) {
            cout << str[i] << " ";
        }
        cout << endl;
        exit(0);
    }

    rep(i, tree[index].size()) {
        ll hoge = tree[index][i];
        // 同じ道は通らない
        if (dic.find(hoge) != dic.end()) continue;

        dfs(hoge, y, route, tree);
    }

    route.pop();
    return route;
}

int main() {
    ll n, x, y;
    cin >> n >> x >> y;
    x -= 1;
    y -= 1;
    // cout << n << x << y << endl;
    vvll tree(n);
    for (int i = 0; i < n - 1; ++i) {
        ll u, v;
        // cout << " scanf ";
        //        scanf("%lld %lld", &u, &v);
        cin >> u >> v;
        u -= 1;
        v -= 1;

        // cout << "i:" << i << endl;

        tree[u].emplace_back(v);
        tree[v].emplace_back(u);
    }

    stack<ll> route;
    dfs(x, y, route, tree);

    return 0;
}